Tensile/KernelWriterAssembly.py

```diff
@@ -12221,54 +12279,34 @@ def globalWriteBatch(self, kernel, ss, batchIdx, applyAlpha, beta, edge, atomic,
-          vgprIdx = 1*(bpm//4)   # index register
 
-          for avi in range(0, gwvw//atomicW):
-            dataV = ss.elementData[elementIdx] + int(avi*ss.cfg.numVgprsPerDataPerVI)
-            atomicDestVgpr = dataV if kernel["BufferStore"] else dataV+2
-            sumIdxV = ss.elementSumIdx[elementIdx] + avi
-            if kernel["ProblemType"]["DestDataType"].numRegisters() < 1 and not kernel["_GlobalAccumulation"]:
-              sumIdxV //= 2
-            if kernel["ProblemType"]["DestDataType"].isDouble():  sumIdxV =  sumIdxV * 2
-
-            # apply mask for element
-            kStr += inst("s_mov_b{}".format(wavelen), self.exec, sgpr(mask,laneSGPRC), "must try again" )
-            if kernel["ProblemType"]["DestDataType"].isDouble():
-              #64-bit C val move by 2 32-bit instructions
-              kStr += inst("v_mov_b32", vgpr(dataV+2), vgpr(atomicDestVgpr), "dataV+2 = tmp (new original C)" )
-              kStr += inst("v_mov_b32", vgpr(dataV+3), vgpr(atomicDestVgpr+1), "dataV+3 = tmp (new original C)" )
-            else:
-              kStr += inst("v_mov_b32", vgpr(dataV+1), vgpr(atomicDestVgpr), "dataV+1 = tmp (new original C)" )
-            kStr += self.chooseAddForAtomic(kernel, \
-                      vgpr(dataV+0,vgprCnt), vgpr(dataV+1*vgprIdx,vgprCnt), vgpr("ValuC+%u"%sumIdxV,vgprCnt), \
-                      "newC = rC + originalC")
-            if self.do["GlobalWrite"]:
-              if kernel["BufferStore"]:
-                # Using no-ret version here?
-                # cmpswap_x2 for DGEMM
-                if kernel["ProblemType"]["DestDataType"].isDouble():
-                  kStr += "_buffer_atomic_cmpswap_b64 %s, %s, %s %s    // %s%s" % \
-                    (vgpr(dataV,4), \
-                     vgpr(addr,1), \
-                     sgpr("SrdD", 4), \
-                     "0 offen offset:%u glc" % (addrCalc.globalOffset), \
-                     "try again", self.endLine )
-                else:
-                  kStr += "_buffer_atomic_cmpswap_b32 %s, %s, %s %s    // %s%s" % \
-                      (vgpr(dataV,2), \
-                       vgpr(addr,1), \
-                       sgpr("SrdD", 4), \
-                       "0 offen offset:%u glc" % (addrCalc.globalOffset), \
-                       "try again", self.endLine )
-              else:
-                kStr += "_global_atomic_cmpswap_b32 %s, %s, %s, %s %s    // %s%s" % ( vgpr(atomicDestVgpr), \
-                    vgpr(addr,2), vgpr(dataV,2), "off", "glc", "try again", self.endLine)
+          loopCount = 0
+          for avi in range(0, gwvw, atomicW):
+            for compIdx in range(vgprLoadDW // atomicOpW):
+              maskIdx = mask + loopCount * laneSGPRC
+              compOffset = compIdx * atomicOpW
+              compDstRegOffset = compOffset * 2
+              dataV = ss.elementData[elementIdx] + int(avi*ss.cfg.numVgprsPerDataPerVI)
+              sumIdxV = ss.elementSumIdx[elementIdx] + avi
```

Comment:
Is avi correct value for this calculation? Was previously a smaller number: gwvw/atomicW vs gwvw

Reply:
I would say the previous one was incorrect.
I added a lot of test cases and they all passed with the current implementation.